pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -332,6 +340,31 @@ private void validateRequest(BrokerRequest brokerRequest) {
             "Value for 'LIMIT' (" + limit + ") exceeds maximum allowed value of " + _queryResponseLimit);
       }
     }
+
+    // Checks whether the query contains non-existent columns.
+    // Table name has already been verified before hitting this line.
+    if (requestParams.isValidateQuery()) {
+      String tableName = brokerRequest.getQuerySource().getTableName();
+      Schema schema = _tableSchemaCache.getIfTableSchemaPresent(tableName);
+      if (schema != null) {
+        RequestInfo requestInfo = RequestUtils.preComputeRequestInfo(brokerRequest);
+        // gets all the columns from broker request
+        Set<String> columnsFromBrokerRequest = requestInfo.getAllColumns();
```

Comment:
Consider adding getPhysicalColumns so the virtual column filtering can be done internally.